### PR TITLE
ANI pre-computation 

### PR DESCRIPTION
**Pull request by maxentile**
## Description

A performance bottleneck in our experiments with re-fitting ANI to free energy data has been to repeatedly compute dU/dtheta on a large collection of static snapshots x that doesn't change from iteration to iteration, where theta is a small subset of model parameters that governs only the final layer in a neural network inside the definition of U. A lot of expensive parts of that computation will be identical across all iterations. Can we split those out?

In this PR, we re-write U(x, theta) = g(f(x), theta), where f doesn't depend on the parameters theta we are varying, and where g is much cheaper than evaluating U from scratch.

Specifically, we define f to be all of the steps up until the last neural network layer (computing symmetry functions, then computing all neural network layers up until the last one), and we define g to be the very last layer.

Precomputing y = f(x) and only re-computing g(y) could potentially speed up our free-energy-fitting parameter optimizations quite a bit. The cost of recomputing U from scratch on a batch of 1000 methane configurations is over 1 second, but the cost of computing just the last layer g is ~40 milliseconds.

```
coordinates.shape torch.Size([1000, 5, 3])
time to compute energies in batch: 1.227 s
time to compute (dU / d all_params) in batch: 1.353 s
energy error: |g(f(x)) - model(x)|:  0.0
time to precompute up until last layer (f): 1.271 s
time to compute last layer (g): 0.038 s
time to compute derivatives of E w.r.t. last layer: 0.045 s
```

(Those timings were measured on a CPU instance in this Colab notebook: [Open In Colab](https://colab.research.google.com/drive/18N7quYsraPNpCS-JYUZxRwyil8NFbv7_?usp=sharing) )

(Thanks to @jaimergp and @wiederm for a helpful discussion this week that led me to try this.)

## Questions
- [x] Should we generalize this to "checkpoint" somewhere else in the computation? (For our purposes, is it safe to assume we only ever want to touch the last layer?)
- [x] Here I hardcoded the assumption that the last layer dimension is 160 -- this is true for the ANI-2x networks for carbon and hydrogen, but invalid for the networks for the other elements (for nitrogen and oxygen networks the last layer dimension is 128, and for sulfur, fluorine, and chlorine the last layer dimension is 96). How best to handle this "jagged" computation? (Two options come to mind: (1) emulate jagged array using sufficiently large padded array, (2) instead of returning/accepting a single array in intermediate computation, return a dictionary mapping element symbols to possibly different-sized arrays. I'm leaning towards option (2).)
- [ ] If further speed-up is needed, does g play nice with torchscript?
- [ ] This seems like something other PyTorchers must have wanted to do in other transfer-learning-y settings -- is there a more concise and maintainable way to do this checkpointing, that doesn't require reaching so far into the guts of the model?

## Status
- [x] Working for ANI-2x element-specific neural networks with last-layer dimension=160 (C, H)
- [x] Working for all the other elements (N, O, S, F, Cl)
- [x] Performance measured for droplet-sized systems
- [x] Incorporated into AlchemicalANI and friends
- [x] Incorporated into overall free-energy-fitting workflow
- [ ] Louder warning labels affixed to brittle dimension assumptions

### CODE REVIEW

**Comment by wiederm** [comment]
I have changed the base to `dev-rew` since that's the place I have implemented the different improvements for the reweighting runs.
I think it will be easier to immediately add to the more readable way the ANI classes are setup in this branch.

**Comment by wiederm** [comment]
Ah, sorry, just realized that you cloned from master!

**Comment by wiederm** [comment]
This is incredible! The time for a system with 100 snapshots and 50 atoms:
```
TorchANI version:  2.1.1
time to compute energies in batch: 9.698 s
energy error: |g(f(x)) - model(x)|:  tensor(0., grad_fn=<NormBackward0>)
time to precompute up until last layer (f): 9.454 s
time to compute last layer (g): 0.028 s
time to compute derivatives of E w.r.t. last layer: 0.041 s
```

**Comment by jaimergp** [comment]
I don't know if this also applies to the kind of optimizations you are investigating, but I stumbled upon this paragraph [in the PyTorch docs](https://pytorch.org/docs/stable/notes/autograd.html#excluding-subgraphs-from-backward) and might be relevant:

> This [requires_grad] is especially useful when you want to freeze part of your model, or you know in advance that you’re not going to use gradients w.r.t. some parameters. For example if you want to finetune a pretrained CNN, it’s enough to switch the requires_grad flags in the frozen base, and no intermediate buffers will be saved, until the computation gets to the last layer, where the affine transform will use weights that require gradient, and the output of the network will also require them.

**Comment by maxentile** [comment]
Thanks Jaime!

In our case, it is also important to avoid computing the forward pass. (If we set requires_grad=False for everything but the last layer, this will successfully avoid some unnecessary computation for the _backward_ pass, but it does not avoid the expensive computation of the _forward_ pass.) Unfortunately it looks like I didn't save the timing results for this approach when I initially tried this (maybe @wiederm has?), but at best I think this optimization can cut the expense by about a half.

File: scripts/precomputation.py [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+        aev = aev.flatten(0, 1)
+
+        output = aev.new_zeros(species_.shape)
+
+        for i, (_, m) in enumerate(self.items()):
+            mask = species_ == i
+            midx = mask.nonzero().flatten()
+            if midx.shape[0] > 0:
+                input_ = aev.index_select(0, midx)
+                if i == 2:
```
Comment (maxentile) [inline scripts/precomputation.py:106]: Nice! Minor tweak: rather than having an if/else block that covers element index i=2 and i=3, could we have an array (or dict) that maps `i` to `last_layer_dim`? Then we would say something like `input_ = input_[:, :last_layer_dim[i]]`

Comment (wiederm) [reply scripts/precomputation.py:106]: yes, great point! I will update this!

File: scripts/precomputation.py [hunk 2]
```diff
@@ -86,6 +86,15 @@ class LastLayerANIModel(ANIModel):
     """just like ANIModel, but only does the final calculation and cuts input arrays to the input feature size of the
     different atom nets!"""
 
+    last_two_layer_nr_of_feature: dict = {
+        3: {0: 192, 1: 192, 2: 160, 3: 160},
+        1: {0: 160, 1: 160, 2: 128, 3: 128},
+    }
```
Comment (wiederm) [inline scripts/precomputation.py:92]: addressing @maxentile suggestion!

**Comment by wiederm** [comment]
I think we have identified two major causes of the slow computations and have fixed them here.
One of the issues were the restraints. To show some timing results:
for a droplet with 55 atoms and 88 frames including restraints (all results generated with the `timing.py` script uploaded in the previous commit:
`time to compute energies in batch: 18.115 s`
and excluding restraints:
`time to compute energies in batch: 9.577 s`
For postprocessing we don't need the restraints, so we can remove them.

Also, we don't need to recompute the results for all layers in the reweighting scheme, something 
that was addressed above by @maxentile . The timing below is with the same system as above:
```
time to compute energies in batch: 9.132 s
First time to calculate energies with CompartimentedAlchemicalANI2x: 9.693 s
Second time to calculate energies with CompartimentedAlchemicalANI2x: 0.303 s
```
That means we are now down to < 1s. 
This timing will for sure increase when we include it in the training scheme since we are setting up a few things for each tautomer pair, but not by much. Even if it increases to 10s this is absolutely fine.

I will report the timing for the training loops next.

**Comment by jchodera** [comment]
Any idea why restraints double the computation time?

**Comment by wiederm** [comment]
The main issue here is that we loop through all frames and calculate the multiple restraints individually. It seems as if calculating the restraint for the system takes <0.01 s, but that is enough that in a loop over 800 frames it accumulates to 8s.

**Comment by wiederm** [comment]
The timing for a training run with 3 epochs for a single tautomer pair in a droplet using 1100 frames and 60 atoms has now changed from  ~600 s to 20 s.

One thing that has increased is the hard disk footprint. We are now pre-generating a lot of calculations and we are caching all of this on disk. Depending on the size of the system and the number of frames included this can get very big.

E.g. for ~200 atoms and 2000 frames it takes ~4GB of hard disk for a single tautomer pair. I will investigate this more, but currently I am willing to pay this price for the massively increased speed.

**Comment by wiederm** [comment]
I should add here that there is quiet a big overhead for setting all of this up in the beginning. The timing for a single gradient update per molecule is 2s.

**Comment by maxentile** [comment]
Awesome work! The main things I'll focus on in my review this evening / weekend, based on discussion with @wiederm :
- [ ] The new mixins and class hierarchy
- [ ] Tests that are failing in PyTest but passing when run individually
- [ ] Possibility for I/O performance improvements (compressed pickle vs. numpy.save)
- [ ] Any steps that could be error-prone due to in-place modification of PyTorch tensors
- [ ] Any steps that could be error-prone in the jagged array emulation
